### PR TITLE
FM2-366: Implement Encounter.period mapping

### PR DESCRIPTION
**Pull request by pmanko**
## Description of what I changed
Adding the functionality to map between:
- OpenMRS Encounter `encounterDatetime` and FHIR `Encounter.period`
- OpenMRS Visit `startDatetime` and `stopDatetime` and FHIR `Encounter.period`

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-366

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/VisitTranslatorImplTest.java [hunk 1]
```diff
@@ -290,4 +333,25 @@ public void toFhirResource_shouldTranslateTypeToEncounterType() {
...
+		period.setStart(periodStart);
+		period.setEnd(periodEnd);
+		
+		visit.setStartDatetime(periodStart);
+		visit.setStartDatetime(periodEnd);
+		
+		when(visitPeriodTranslator.toFhirResource(ArgumentMatchers.any())).thenReturn(period);
+		
+		Encounter result = visitTranslator.toFhirResource(visit);
+
```
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/VisitTranslatorImplTest.java:352]: @pmanko  thanks for working on this,add a ticket id to your commit and you can also remove the spaces from your code

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/EncounterPeriodTranslatorImpl.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.api.translators.impl;
+
+import javax.annotation.Nonnull;
+
```
Comment (HerbertYiga) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/EncounterPeriodTranslatorImpl.java:13]: remove space

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/EncounterPeriodTranslator.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.fhir2.api.translators;
+
+import javax.annotation.Nonnull;
+
+import java.util.Date;
+
```
Comment (HerbertYiga) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/EncounterPeriodTranslator.java:15]: remove spaces

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/VisitPeriodTranslatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@Setter(AccessLevel.PACKAGE)
+public class VisitPeriodTranslatorImpl implements VisitPeriodTranslator {
+	
+	@Override
+	public Period toFhirResource(@Nonnull ImmutablePair<Date, Date> pair) {
+		Period period = new Period();
+
```
Comment (HerbertYiga) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/VisitPeriodTranslatorImpl.java:30]: remove space

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/EncounterPeriodTranslatorImplTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	public void setup() {
+		encounterPeriodTranslator = new EncounterPeriodTranslatorImpl();
+	}
+	
+	@Test
+	public void toFhirResource_shouldMapEncounterDatetimeToPeriod() {
+		Date encounterDate = new java.util.Date();
+		
+		Period result = encounterPeriodTranslator.toFhirResource(encounterDate);
+
```
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/EncounterPeriodTranslatorImplTest.java:40]: remove space

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/EncounterPeriodTranslatorImplTest.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	public void toOpenmrsObject_shouldMapPeriodToEncounterDatetime() {
+		Date encounterDate = new java.util.Date();
+		
+		Period period = new Period();
+		period.setStart(encounterDate);
+		
+		Date result = encounterPeriodTranslator.toOpenmrsType(period);
+		
+		assertThat(result, notNullValue());
+		assertThat(result, equalTo(encounterDate));
```
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/EncounterPeriodTranslatorImplTest.java:57]: thanks i was looking for this line ie testing that actually the dates are equal

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/VisitTranslatorImplTest.java [hunk 7]
```diff
@@ -128,6 +133,44 @@ public void toOpenmrsType_shouldConvertIdToUuid() {
...
+		period.setStart(periodStart);
+		period.setEnd(periodEnd);
+		encounter.setPeriod(period);
+		
+		ImmutablePair<Date, Date> pair = new ImmutablePair(periodStart, periodEnd);
+		
+		when(visitPeriodTranslator.toOpenmrsType(period)).thenReturn(pair);
+		
+		Visit result = visitTranslator.toOpenmrsType(new Visit(), encounter);
+
```
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/VisitTranslatorImplTest.java:150]: remove space

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/EncounterPeriodTranslator.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import org.hl7.fhir.r4.model.Period;
+
+public interface EncounterPeriodTranslator extends ToFhirTranslator<Date, Period>, OpenmrsFhirTranslator<Date, Period> {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/EncounterPeriodTranslator.java:18]: Wouldn't it be better to structure this as an interface that translates between an OpenMRS type and a date? That way, we only need a single interface (`EncounterPeriodTranslator`) that has two specialisations (one for `Encounter`, one for `Visit`) and we don't have to do weird things like returning `Pairs`

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/EncounterTranslatorImplTest.java [hunk 9]
```diff
@@ -22,12 +22,7 @@
 import java.util.Set;
 
 import org.exparity.hamcrest.date.DateMatchers;
-import org.hl7.fhir.r4.model.CodeableConcept;
-import org.hl7.fhir.r4.model.Encounter;
-import org.hl7.fhir.r4.model.IdType;
-import org.hl7.fhir.r4.model.Provenance;
-import org.hl7.fhir.r4.model.Reference;
-import org.hl7.fhir.r4.model.Resource;
+import org.hl7.fhir.r4.model.*;
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/EncounterTranslatorImplTest.java:25]: Please don't use `*` imports. This is an OpenMRS convention.

File: omod/pom.xml [hunk 10]
```diff
@@ -132,5 +132,37 @@
...
+									<version>${project.version}</version>
+									<groupId>${parent.groupId}</groupId>
+								</configuration>
+							</execution>
+						</executions>
+					</plugin>
+				</plugins>
+			</build>
+		</profile>
+	</profiles>
```
Comment (ibacher) [inline omod/pom.xml:166]: This is maybe something we should discuss separately from this PR?

File: pom.xml [hunk 11]
```diff
@@ -766,5 +766,56 @@
...
+                        <executions>
+                            <execution>
+                                <id>validate</id>
+                                <phase>none</phase>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
```
Comment (ibacher) [inline pom.xml:819]: Ditto. PS I think I separately disabled `test-jar`s in [this commit](https://github.com/openmrs/openmrs-module-fhir2/commit/bd57419d45feadab98554b338eb4b3b8c0fdeba2).

**Review by ibacher — CHANGES_REQUESTED** [review]

**Comment by ibacher** [comment]
@pmanko Any update on this?

**Review by ibacher — APPROVED** [review]

**Pull request merged**